### PR TITLE
Publish Docker image to any host

### LINKED ISSUE #46: Docker publish action should support login-server
This would allow us to push to hosts other than Docker Hub. Ideally we can use it to push to Azure https://github.com/Azure/docker-login

### PR DESCRIPTION
**Pull request by MichaelKora**
Fixes #46

### CODE REVIEW

File: actions/docker-publish/action.yaml [hunk 1]
```diff
@@ -1,60 +1,61 @@
...
-          exit 1
-        fi
+    # setup-buildx action will create and boot a builder using by default the docker-container driver.
+    # This is not required but recommended using it to be able to build multi-platform images, export cache, etc.
+    - name: Set up Docker Buildx
+      uses: docker/setup-buildx-action@v2.2.1
 
-        docker load --input build/*.tar
+    - name: Login to the Registry
+      uses: "docker/login-action@v2.1.0"
```
Comment (disrupted) [inline actions/docker-publish/action.yaml:48]: cool! 👍🏻 I guess this is even better than the azure login action which is mentioned in the linked issue, because it supports multiple Docker registries, right?

File: actions/docker-publish/action.yaml [hunk 2]
```diff
@@ -1,60 +1,61 @@
...
-      shell: bash
-      working-directory: ${{ inputs.working-directory }}
+    - name: Build and push
+      uses: docker/build-push-action@v3.2.0
+      with:
+        context: "${{ inputs.working-directory }}"
+        push: true
+        tags: |
+          ${{ inputs.docker-registry }}/${{ inputs.publisher }}/${{ inputs.image-name }}:${{ inputs.image-tag }}
+          ${{ inputs.docker-registry }}/${{ inputs.publisher }}/${{ inputs.image-name }}:latest
```
Comment (disrupted) [inline actions/docker-publish/action.yaml:61]: Why did you change the image tag logic? This is a big breaking change and shouldn't be done unless it is specifically required. Especially we wouldn't want to publish `latest` accidentally for example

**Review by disrupted — CHANGES_REQUESTED** [review]

File: actions/docker-publish/action.yaml (hunk 2 above)
Comment (MichaelKora) [reply actions/docker-publish/action.yaml:61]: i am using it in [another CI](https://github.com/bakdata/cluster-shell/pull/1) where i release a new tag and then use the new tag to build and publish a new docker image and also update the index.yaml on `gh-pages`.

File: actions/docker-publish/action.yaml [hunk 3]
```diff
@@ -1,60 +1,61 @@
...
-          docker push ${{ inputs.publisher }}/${{ inputs.image-name }}:latest
-          docker push ${{ inputs.publisher }}/${{ inputs.image-name }}:${GITHUB_REF/refs\/tags\//}
-        else
-          docker tag ${{ inputs.image-name }}:${{ github.run_id }} ${{ inputs.publisher }}/${{ inputs.image-name }}:${{ github.run_id }}
-          docker push ${{ inputs.publisher }}/${{ inputs.image-name }}:${{ github.run_id }}
-        fi
-      shell: bash
-      working-directory: ${{ inputs.working-directory }}
+    - name: Build and push
+      uses: docker/build-push-action@v3.2.0
```
Comment (yannick-roeder) [inline actions/docker-publish/action.yaml:55]: What about publishing a Docker image that was build with [Jib](https://github.com/GoogleContainerTools/jib) from Java source code? This action should only publish Docker images because we need to support all use cases and building a Docker image from a Dockerfile can be provided as a workflow template if needed

Comment (yannick-roeder) [reply actions/docker-publish/action.yaml:55]: This was also the reason for allowing a `.tar` file as an input

File: actions/docker-publish/action.yaml (hunk 2 above)
Comment (yannick-roeder) [reply actions/docker-publish/action.yaml:61]: How do you support pushing dev images from feature or fix branches like this? Developers need to be able to test there image before merging without overwriting the latest tag

File: actions/docker-build/README.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+| ------------------- | :------: | :--------------------------: | :----: | ----------------------------------------------------------------------------------------------------------------------------------------------------- |
+| dockerfile-dir      |    ✅    |              -               | string | Directory contining the dockerfile                                                                                                                    |
+| github-token        |    ✅    |              -               | string | Github token to use for checkout.                                                                                                                     |
+| image-name          |    ✅    | github.event.repository.name | string | Name of Docker image on Dockerhub                                                                                                                     |
+| image-artifact-name |    ❌    |              -               | string | PuName of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact'). |
```
Comment (disrupted) [inline actions/docker-build/README.md:16]:
```suggestion
| image-artifact-name |    ❌    |              -               | string | Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact'). |
```

**Review by disrupted — CHANGES_REQUESTED** [review]

File: actions/docker-build/action.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+name: "Publish Docker image"
+description: "Use a Dockerfile to publish image to any host"
```
Comment (yannick-roeder) [inline actions/docker-build/action.yaml:2]: Please change

File: actions/docker-build/action.yaml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+inputs:
+  dockerfile-dir:
+    description: "Directory contining the dockerfile"
```
Comment (yannick-roeder) [inline actions/docker-build/action.yaml:6]:
```suggestion
    description: "Directory containing the Dockerfile"
```

File: actions/docker-build/action.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    required: true
+  github-token:
+    description: "Github token."
```
Comment (yannick-roeder) [inline actions/docker-build/action.yaml:9]:
```suggestion
    description: "GitHub token"
```

File: actions/docker-publish/action.yaml [hunk 8]
```diff
@@ -1,24 +1,34 @@
 name: "Publish Docker image"
-description: "Publish Docker image from download artifact"
+description: "Publish Docker image from download artifact to any host"
 
 inputs:
-  image-artifact-name:
-    description: "Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact')."
+  docker-registry:
+    description: "Host where the image should be pushed to."
+    required: true
```
Comment (yannick-roeder) [inline actions/docker-publish/action.yaml:7]: This should not be required and should be DockerHub by default

File: actions/docker-publish/action.yaml [hunk 9]
```diff
@@ -1,24 +1,34 @@
...
+  github-token:
+    description: "Github token."
```
Comment (yannick-roeder) [inline actions/docker-publish/action.yaml:9]:
```suggestion
    description: "GitHub token"
```

File: actions/docker-build/action.yaml (hunk 5 above)
Comment (yannick-roeder) [reply actions/docker-build/action.yaml]: Please also change the name

File: actions/docker-build/action.yaml [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    required: false
+    default: ${{ github.ref_name }}
+runs:
+  using: "composite"
+  steps:
+    - name: Check out repository
+      uses: actions/checkout@v3
+      with:
+        ref: ${{ inputs.ref }}
+        token: ${{ inputs.github-token }}
```
Comment (yannick-roeder) [inline actions/docker-build/action.yaml:30]: Why do you need a token here to checkout the repository? We don't need it in other actions or workflows

File: actions/docker-build/action.yaml [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+    - name: Build Dockerfile and store them in "build" directory
+      run: |
+        docker build -t ${{ inputs.image-name }}:${{ github.run_id }} ${{ inputs.dockerfile-dir }}/
```
Comment (yannick-roeder) [inline actions/docker-build/action.yaml:34]: Did you test this with setting `dockerfile-dir` to something else than `.`? I'm pretty sure from previous testing that you also need a `--file` flag additionally for other paths to work

File: .github/workflows/docker-build-and-publish.yaml [hunk 12]
```diff
@@ -0,0 +1,77 @@
+name: Docker build and publish
+# Reusable workflow for using a dockerfile to build and push image to any container registry
```
Comment (yannick-roeder) [inline .github/workflows/docker-build-and-publish.yaml:2]:
```suggestion
# Reusable workflow to build and push an image from a Dockerfile to any container registry
```

File: .github/workflows/docker-build-and-publish.yaml [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+        required: true
+
+jobs:
+  docker-build:
+    name: Docker build
+    runs-on: ubuntu-22.04
+
+    steps:
+      - name: Build docker image and upload tar file
+        uses: bakdata/ci-templates/actions/docker-build@feat/docker-publish
```
Comment (yannick-roeder) [inline .github/workflows/docker-build-and-publish.yaml:55]: ToDo: bump before merge

File: .github/workflows/docker-build-and-publish.yaml [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+          github-token: "${{ secrets.github-token }}"
+          dockerfile-dir: "${{ inputs.dockerfile-dir }}"
+
+  docker-publish:
+    name: Docker publish
+    runs-on: ubuntu-22.04
+    needs: [docker-build]
+    steps:
+      - name: Use uploaded tar files to push new image to docker
+        uses: bakdata/ci-templates/actions/docker-publish@feat/docker-publish
```
Comment (yannick-roeder) [inline .github/workflows/docker-build-and-publish.yaml:68]: ToDo: bump before merge

File: .github/workflows/README.md [hunk 15]
```diff
@@ -12,6 +13,72 @@ The following workflows can be found here:
...
+This workflow requires a Dockerfile located in the repository.
+
+### Dependencies
+
+This workflow is built from multiple composite actions listed below:
+
+- [docker-build](https://github.com/bakdata/ci-templates/tree/main/actions/docker-build)
+- [docker-publish](https://github.com/bakdata/ci-templates/tree/main/actions/docker-publish)
+- [helm-package](https://github.com/bakdata/ci-templates/tree/main/actions/helm-package)
+- [commit-and-push](https://github.com/bakdata/ci-templates/tree/main/actions/commit-and-push)
```
Comment (yannick-roeder) [inline .github/workflows/README.md:31]: Where are you using the `helm-package` and `commit-and-push` actions in your workflow?

File: .github/workflows/README.md [hunk 16]
```diff
@@ -12,6 +13,72 @@ The following workflows can be found here:
...
+
+```yaml
+name: Docker build and publish
+
+on:
+  workflow_dispatch:
+
+jobs:
+  call-workflow-passing-data:
+    name: Publish Helm chart
```
Comment (yannick-roeder) [inline .github/workflows/README.md:65]:
```suggestion
    name: Build and push Docker image
```

File: .github/workflows/README.md [hunk 17]
```diff
@@ -12,6 +13,72 @@ The following workflows can be found here:
...
+    uses: bakdata/ci-templates/.github/workflows/docker-build-and-publish.yaml@main
+    with:
+      dockerfile-dir: "./kafka-client"
```
Comment (yannick-roeder) [inline .github/workflows/README.md:68]: Please be less specific, like `./my-docker-dir` or similar

File: .github/workflows/README.md [hunk 18]
```diff
@@ -12,6 +13,72 @@ The following workflows can be found here:
...
+      docker-registry: "my-registry.com"
+      image-name: "kafka-client"
```
Comment (yannick-roeder) [inline .github/workflows/README.md:70]: Please be less specific, like `my-image` or similar

**Review by yannick-roeder — CHANGES_REQUESTED** [review]

File: actions/docker-publish/action.yaml [hunk 19]
```diff
@@ -32,29 +42,35 @@ runs:
         name: ${{ inputs.image-artifact-name }}
         path: build
 
-    - name: Login to DockerHub
-      uses: docker/login-action@v1
+    # setup-buildx action will create and boot a builder using by default the docker-container driver.
+    # This is not required but recommended using it to be able to build multi-platform images, export cache, etc.
```
[the remark this answers is not in the document]
Comment (disrupted) [reply actions/docker-publish/action.yaml:46]: https://github.com/marketplace/actions/docker-setup-buildx

File: .github/workflows/README.md [hunk 20]
```diff
@@ -12,6 +13,72 @@ The following workflows can be found here:
...
+
+This workflow requires a Dockerfile located in the repository.
+
+### Dependencies
+
+This workflow is built from multiple composite actions listed below:
+
+- [docker-build](https://github.com/bakdata/ci-templates/tree/main/actions/docker-build)
+- [docker-publish](https://github.com/bakdata/ci-templates/tree/main/actions/docker-publish)
+- [helm-package](https://github.com/bakdata/ci-templates/tree/main/actions/helm-package)
```
Comment (disrupted) [inline .github/workflows/README.md:30]: why does the Docker workflow use Helm?

File: .github/workflows/README.md [hunk 21]
```diff
@@ -50,8 +117,6 @@ This workflow is built from multiple composite actions listed below:
 
 - [helm-lint](https://github.com/bakdata/ci-templates/tree/main/actions/helm-lint)
 - [bump-version](https://github.com/bakdata/ci-templates/tree/main/actions/bump-version)
-- [helm-package](https://github.com/bakdata/ci-templates/tree/main/actions/helm-package)
-- [commit-and-push](https://github.com/bakdata/ci-templates/tree/main/actions/commit-and-push)
```
Comment (disrupted) [inline .github/workflows/README.md:54]: why did you remove those?

File: .github/workflows/README.md [hunk 22]
```diff
@@ -12,6 +13,72 @@ The following workflows can be found here:
...
+| ------------------- | :------: | :--------------------------------: | :----: | --------------------------------------------------------------------------------------------------------------------------------------------------- |
+| dockerfile-path     |    ❌    |                 -                  | string | Path to the Dockerfile.                                                                                                                             |
+| docker-registry     |    ❌    |          "hub.docker.com"          | string | Host where the image should be pushed to.                                                                                                           |
+| image-artifact-name |    ❌    |          "image-artifact"          | string | Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact'). |
+| image-name          |    ❌    | ${{ github.event.repository.name}} | string | Name of Docker image (Default is the repository name).                                                                                              |
```
Comment (disrupted) [inline .github/workflows/README.md:40]:
```suggestion
| image-name          |    ❌    | ${{ github.event.repository.name }} | string | Name of Docker image (Default is the repository name).                                                                                              |
```

File: actions/docker-build/README.md (hunk 4 above)
Comment (disrupted) [reply actions/docker-build/README.md:15]: this is not resolved, it still shows ups for me in the diff

**Review by disrupted — CHANGES_REQUESTED** [review]

File: .github/workflows/README.md (hunk 21 above)
Comment (MichaelKora) [reply .github/workflows/README.md:54]: because none of them is used in `Docker Build and Publish`

File: actions/docker-build/action.yaml [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    default: "./Dockerfile"
+  github-token:
+    description: "GitHub token"
+    required: true
+  image-artifact-name:
+    description: "Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact')."
+    required: false
+    default: "image-artifact"
+  image-name:
+    description: "Name of Docker image (Default is the repository name)."
```
Comment (disrupted) [inline actions/docker-build/action.yaml:17]: where do you set the default?

File: actions/docker-build/action.yaml (hunk 10 above)
Comment (yannick-roeder) [reply actions/docker-build/action.yaml:31]: But this action is not creating the tag right? I don't mind optionally allowing the user to set a `github-token`, but it shouldn't be required then since this action can also be used directly and on other branches

File: actions/docker-build/README.md [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+| dockerfile-path     |    ✅    |              -               | string | Path to the Dockerfile.                                                                                                                             |
+| github-token        |    ✅    |              -               | string | Github token to use for checkout.                                                                                                                   |
+| image-artifact-name |    ❌    |              -               | string | Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact'). |
+| image-name          |    ❌    | github.event.repository.name | string | Name of Docker image on Dockerhub                                                                                                                   |
+| ref                 |    ❌    |       github.ref_name        | string | Branch to use for the checkout.                                                                                                                     |
```
Comment (disrupted) [inline actions/docker-build/README.md:17]: I believe action parameters don't have types

File: actions/docker-publish/README.md [hunk 25]
```diff
@@ -1,21 +1,23 @@
...
+| docker-registry     |    ❌    |       "hub.docker.com"       | string | Host where the image should be pushed to.                                                                                                           |
+| image-artifact-name |    ❌    |       "image-artifact"       | string | Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact'). |
+| image-name          |    ❌    | github.event.repository.name | string | Name of Docker image on Dockerhub                                                                                                                   |
+| ref                 |    ❌    |       github.ref_name        | string | Branch to use for the checkout.                                                                                                                     |
+| working-directory   |    ❌    |              -               | string | Working directory for your Docker artifacts                                                                                                         |
```
Comment (disrupted) [inline actions/docker-publish/README.md:20]: only workflows parameters have types

File: actions/docker-build/action.yaml (hunk 5 above)
Comment (yannick-roeder) [reply actions/docker-build/action.yaml]: This is not resolved. The name is still `Publish Docker image`, but this action is not publishing anything

**Review by disrupted — DISMISSED** [review]

File: actions/docker-publish/README.md [hunk 26]
```diff
@@ -24,10 +26,12 @@ steps:
...
-      username: ${{ secrets.docker-user }}
-      password: ${{ secrets.docker-password }}
-      working-directory: "." # (Optional)
+      image-tag: "v.1.0"
+      image-name: "tarball"
+      publisher: "my-publisher"
+      username: "${{ secrets.docker-user }}"
+      password: "${{ secrets.docker-password }}"
+      working-directory: "./tarball"
+      github-token: "${{ secrets.GITHUB_TOKEN }}"
```
[the remark this answers is not in the document]
Comment (yannick-roeder) [reply actions/docker-publish/README.md:35]: Actions don't support secrets, only workflows do AFAIK

[the remark this answers is not in the document]
Comment (disrupted) [reply actions/docker-publish/README.md:35]: if GitHub intends it to be used that way, I assume it should be fine. I certainly don't know of any alternatives

File: .github/workflows/README.md [hunk 27]
```diff
@@ -12,6 +13,71 @@ The following workflows can be found here:
...
+| docker-context      |    ❌    |                 "."                 | string | The docker context.                                                                                                                                 |
+| dockerfile-path     |    ❌    |            "Dockerfile"             | string | Path to the Dockerfile.                                                                                                                             |
+| docker-registry     |    ❌    |                 ""                  | string | Host where the image should be pushed to.                                                                                                           |
+| image-artifact-name |    ❌    |          "image-artifact"           | string | Name of the artifact that contains the Docker image.tar file to push, see https://github.com/actions/upload-artifact (Default is 'image-artifact'). |
+| image-name          |    ❌    | ${{ github.event.repository.name }} | string | Name of Docker image (Default is the repository name).                                                                                              |
```
Comment (yannick-roeder) [inline .github/workflows/README.md:39]: Please write the default without variable braces, just `github.event.repository.name`

File: .github/workflows/lint.yaml [hunk 28]
```diff
@@ -12,6 +12,8 @@ jobs:
         shell: bash
     steps:
       - uses: actions/checkout@v3
+        with:
+          persist-credentials: false
```
Comment (yannick-roeder) [inline .github/workflows/lint.yaml:16]: This should not be part of this MR

File: .github/workflows/docker-build-and-publish.yaml [hunk 29]
```diff
@@ -0,0 +1,86 @@
...
+      docker-publisher:
+        description: "Publisher to prefix Docker image (e.g. 'my-publisher')."
+        required: true
+      docker-user:
+        description: "Username for the Docker registry login."
+        required: true
+      docker-password:
+        description: "Password for the Docker registry login."
+        required: true
+      # if this workflow should be automatically triggered after a new tag was pushed by another action, this action will not be triggered. Configure the github token to overcome it (as described here: https://github.com/orgs/community/discussions/27028)
```
Comment (yannick-roeder) [inline .github/workflows/docker-build-and-publish.yaml:51]:
```suggestion
      # if this workflow should be automatically triggered after a new tag was pushed by another action, then configure the github token to overcome it (as described here: https://github.com/orgs/community/discussions/27028)
```

File: .github/workflows/docker-build-and-publish.yaml [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+      github-token:
+        description: "GitHub token."
+        required: true
```
Comment (yannick-roeder) [inline .github/workflows/docker-build-and-publish.yaml:54]: This should not be required. Please also update the docs accordingly

File: .github/workflows/README.md [hunk 31]
```diff
@@ -12,6 +13,71 @@ The following workflows can be found here:
...
+### Secret Parameters
+
+These secrets define the user that pushes the built images to the container registry.
+
+| Name             | Required | Description                                             |
+| ---------------- | :------: | ------------------------------------------------------- |
+| docker-publisher |    ✅    | Publisher to prefix Docker image (e.g. 'my-publisher'). |
+| docker-user      |    ✅    | Username for the Docker registry login.                 |
+| docker-password  |    ✅    | Password for the Docker registry login.                 |
+| github-token     |    ✅    | The GitHub token for the checkout                       |
```
Comment (yannick-roeder) [inline .github/workflows/README.md:52]: Please update the description of the token to the same as it is in the action:
```
GitHub token to use for checkout (important when this action should be automatically triggerd by another action: https://github.com/orgs/community/discussions/27028)
```

File: actions/docker-build/README.md [hunk 32]
```diff
@@ -0,0 +1,32 @@
...
+```yaml
+steps:
+  - name: Build tarball image
+    uses: bakdata/ci-templates/actions/docker-build@main
+    with:
+      dockerfile-context: "./docker-dir/"
+      dockerfile-path: "docker-dir/Dockerfile"
+      image-artifact-name: "${{ inputs.image-artifact-name }}"
+      image-name: "tarball"
+      github-token: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (yannick-roeder) [inline actions/docker-build/README.md:31]:
```suggestion
      dockerfile-context: "./docker-dir/"
      dockerfile-path: "./path/to/my/Dockerfile"
      image-artifact-name: "my-image-artifact"
      image-name: "my-image"
      github-token: "${{ secrets.GH_TOKEN }}"
```

File: .github/workflows/docker-build-and-publish.yaml [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+      docker-user:
+        description: "Username for the Docker registry login."
+        required: true
+      docker-password:
+        description: "Password for the Docker registry login."
+        required: true
+      # if this workflow should be automatically triggered after a new tag was pushed by another action, then configure the github token to overcome it (as described here: https://github.com/orgs/community/discussions/27028)
+      github-token:
+        description: "GitHub token."
+        required: false
```
Comment (yannick-roeder) [inline .github/workflows/docker-build-and-publish.yaml:54]: This needs a default then as well

Comment (MichaelKora) [reply .github/workflows/docker-build-and-publish.yaml]: i set a default value. but i had to move from the `secrets` block since secrets don't have a `default` property

Comment (yannick-roeder) [reply .github/workflows/docker-build-and-publish.yaml]: Ok, that's not ideal, but we have no other choice here I guess

**Review by yannick-roeder — APPROVED** [review]

File: .github/workflows/docker-build-and-publish.yaml (hunk 33 above)
Comment (yannick-roeder) [reply .github/workflows/docker-build-and-publish.yaml]: Don't forget to also change the `secret.` to a `input.` like the failing workflow is telling you

**Review by yannick-roeder — APPROVED** [review]
